File: code_examples/5-Holding_Data/1-Arrays/4-Array_2D/Python/main.py
#!/usr/bin/env python3
"""
Created by: Mr. Coxall
Created on: Sep 2020
This module uses an array as a parameter
"""


import random
from typing import List


# in python an array is passed by reference to a function
def sum_of_numbers(array_of_numbers: List[int]) -> int:
    """The sum_of_numbers() function calculates the sum of numbers in a 2D list, returns the sum as int."""

    total = 0

    for row_value in array_of_numbers:
        for single_value in row_value:
            total += single_value

    return total


def main() -> None:
    """The main() function just calls other functions, returns None."""

    a_2d_list = []
    sum_answer = 0

    # input
    rows = int(input("How many row would you like: "))
    columns = int(input("How many columns would you like: "))

    for _ in range(0, rows):
        temp_column = []
        for _ in range(0, columns):
            a_random_number = random.randint(0, 9)
            temp_column.append(a_random_number)
            print(f"{a_random_number} ", end="")
        a_2d_list.append(temp_column)
        print("")

    sum_answer = sum_of_numbers(a_2d_list)

    # output
    print(f"\nThe sum of all the numbers is: {sum_answer}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: code_examples/5-Holding_Data/1-Arrays/4-Array_2D/Go/main.go
/**
 * Created by: Mr. Coxall
 * Created on: Sep 2020
 * This program uses a 2D array
 */

package main

import (
	"fmt"
	"math/rand"
)

func sumOfNumbers(arrayOfNumbers [][]int) int {
	total := 0

	for _, row := range arrayOfNumbers {
		for _, value := range row {
			total += value
		}
	}

	return total
}

func main() {
	var a2DList [][]int
	var sumAnswer int

	// input
	var rows, columns int
	fmt.Print("How many rows would you like: ")
	fmt.Scanln(&rows)
	fmt.Print("How many columns would you like: ")
	fmt.Scanln(&columns)

	for i := 0; i < rows; i++ {
		tempColumn := make([]int, columns)
		for j := 0; j < columns; j++ {
			aRandomNumber := rand.Intn(10)
			tempColumn[j] = aRandomNumber
			fmt.Printf("%d ", aRandomNumber)
		}
		a2DList = append(a2DList, tempColumn)
		fmt.Println()
	}

	sumAnswer = sumOfNumbers(a2DList)

	// output
	fmt.Printf("\nThe sum of all the numbers is: %d\n", sumAnswer)

	fmt.Println("\nDone.")
}
